thunder/core/rematerialization.py

```diff
@@ -172,15 +172,46 @@ def apply_rematerialization_for_consumer(
     # Some recomputing_symbols might require producer's inputs, so we need to
     # add them to the consumer's inputs.
     # Probably find_min_cut should have returned this information.
-    all_args = tuple(
-        chain.from_iterable(
-            (x.name for x in bsym.flat_args if isinstance(x, ProxyInterface)) for bsym in new_subsymbols
-        )
-    )
+    all_args = set(chain.from_iterable((x.name for x in bsym.flat_proxy_args) for bsym in new_subsymbols))
+    all_outs = set(chain.from_iterable((x.name for x in bsym.flat_proxy_outs) for bsym in new_subsymbols))
     new_consumer_args += tuple(
-        x for x in producer.args if x.name in all_args and x.name not in (x.name for x in new_consumer_args)
+        x
+        for x in producer.args
+        if x.name in all_args and x.name not in (x.name for x in new_consumer_args) and x.name not in all_outs
     )
 
+    def sort_bsyms_by_dataflow(
+        bsyms: list[BoundSymbolInterface], inps: Sequence[ProxyInterface]
+    ) -> list[BoundSymbolInterface]:
+        from collections import deque
+
+        in_degree = []
+        new_bsyms = []
+        q = deque()
+        consumers = utils.consumers(bsyms, _map_to_numbers=True)
+        inp_names = tuple(x.name for x in inps)
+        for bsym in bsyms:
+            in_degree.append(0)
+            for arg in bsym.flat_proxy_args:
+                if not arg.name in inp_names:
+                    in_degree[-1] += 1
+            if in_degree[-1] == 0:
+                q.append(len(in_degree) - 1)
+        while q:
+            idx = q.popleft()
+            new_bsyms.append(bsyms[idx])
+            for out in bsyms[idx].flat_proxy_outs:
+                cons = consumers.get(out, [])
+                for c_idx in cons:
+                    in_degree[c_idx] -= 1
+                    if in_degree[c_idx] == 0:
+                        q.append(c_idx)
+
+        if len(new_bsyms) != len(bsyms):
+            raise RuntimeError("A cyclic dependency was detected in the list of bound symbols.")
+        return new_bsyms
+
+    new_subsymbols = sort_bsyms_by_dataflow(list(new_subsymbols), new_consumer_args)
```

Comment:
Let's reuse existing code:
```suggestion
    _, leaves = bsym_list_to_dag(list(new_subsymbols))
    new_subsymbols = toposort_bsym_dag(leaves, TOPOSORT_ORDER.BOTTOM_UP)
```
The symbols list here doesn't include unpacking operations that act as producers, so a short patch is needed to support it (please add a test case for this change to `test_bsym_toposort`):
```diff
diff --git a/thunder/core/transforms.py b/thunder/core/transforms.py
index 745143df..042468aa 100644
--- a/thunder/core/transforms.py
+++ b/thunder/core/transforms.py
@@ -140,7 +140,9 @@ def bsym_list_to_dag(
     for bsym_id, node in bsym_id_to_node_map.items():
         has_parents: bool = False
         for inp in node.bsym.flat_proxy_args:
-            producer = producers[inp]
+            producer = producers.get(inp, None)
+            if producer is None:
+                continue
             producer_node = bsym_id_to_node_map[producer]
             parent = bsym_id_to_node_map[producer]
```